src/charm.py

```diff
@@ -2,22 +2,208 @@
+            nssf_url=self._nssf_hostname,
+        )
+        self._push_config_file(
+            content=content,
+        )
+        self._nssf_container.add_layer("nssf", self._nssf_pebble_layer, combine=True)
+        self.unit.status = ActiveStatus()
+
+    def _render_config_file(
+        self,
+        nssf_url: str,
+        sbi_port: int,
+        nrf_url: str,
+    ):
+        """Renders the NSSF config file.
+
+        Args:
+            nssf_url (str): URL of the NSSF.
+            sbi_port (int): NSSF SBi port.
+            nrf_url (str): URL of the NRF.
+        """
+        jinja2_environment = Environment(loader=FileSystemLoader(CONFIG_TEMPLATE_DIR))
+        template = jinja2_environment.get_template(CONFIG_TEMPLATE_NAME)
+        content = template.render(
+            sbi_port=sbi_port,
+            nrf_url=nrf_url,
+            nssf_url=nssf_url,
+        )
+        return content
+
+    def _push_config_file(
+        self,
+        content: str,
+    ) -> None:
+        """Writes the NSSF config file and pushes it to the container.
+
+        Args:
+            content (str): Content of the config file.
+        """
+        self._nssf_container.push(
+            path=f"{CONFIG_DIR}/{CONFIG_FILE_NAME}",
+            source=content,
+            make_dirs=True,
+        )
+        logger.info("Pushed %s config file", CONFIG_FILE_NAME)
+
+    def _relation_created(self, relation_name: str) -> bool:
+        """Returns True if the relation is created, False otherwise.
+
+        Args:
+            relation_name (str): Name of the relation.
+
+        Returns:
+            bool: True if the relation is created, False otherwise.
+        """
+        return bool(self.model.get_relation(relation_name))
+
+    @property
+    def _nssf_pebble_layer(self) -> Layer:
+        """Returns pebble layer for the nssf container.
+
+        Returns:
+            Layer: Pebble Layer
+        """
+        return Layer(
+            {
+                "services": {
+                    self._nssf_service_name: {
+                        "override": "replace",
+                        "startup": "enabled",
+                        "command": f"/free5gc/nssf/nssf --nssfcfg {CONFIG_DIR}/{CONFIG_FILE_NAME}",  # noqa: E501
+                        "environment": self._nssf_environment_variables,
+                    },
+                },
+            }
+        )
+
+    @property
+    def _nssf_environment_variables(self) -> dict:
+        """Returns environment variables for the nssf container.
+
+        Returns:
+            dict: Environment variables.
+        """
+        return {
+            "GOTRACEBACK": "crash",
+            "GRPC_GO_LOG_VERBOSITY_LEVEL": "99",
+            "GRPC_GO_LOG_SEVERITY_LEVEL": "info",
+            "GRPC_TRACE": "all",
+            "GRPC_VERBOSITY": "DEBUG",
+            "POD_IP": self._pod_ip,
+            "MANAGED_BY_CONFIG_POD": "true",
+        }
+
+    @property
```

Comment:
Probably shouldn't be a property

Reply:
Why?

Reply:
Well is it really an attribute of the charm? It feels like something we get from outside of the scope of the charm. Just the fact that we can't "set" it makes me think that it shouldn't be a property.

I'm to blame on this but overall it seems like we've overused properties a little bit.

Reply:
I think it can be seen as an attribute of the charm. In this case, the property has not setter, so it is a read-only property, and those are used all the time in Python.

I agree the value we get is not something inside the class or the instance, but I do not think it should prevent the use of property.

I think it makes those the intent of those kind of small lookup methods a bit clearer on use:

```
pod_ip = self._pod_ip

# vs

pod_ip = self._pod_ip()  # to me, this feels like it could be anything, and returning a value
```

I have tried to find opinions on that online, and I have not found anything. In the end, it is a super small detail, and I prefer their use, but I would be open to change the team's guidelines on this.